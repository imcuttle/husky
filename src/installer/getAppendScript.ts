import * as fs from 'fs'
import * as os from 'os'
import * as path from 'path'
import * as slash from 'slash'
import { huskyIdentifier } from './getScript'

interface IContext {
  node: string
  platform: string
  script: string
  version: string
}

// Used to identify scripts created by Husky
export const huskyAppendIdentifier = `${huskyIdentifier}-append`

// Render script
const render = ({ node, platform, script, version }: IContext) => `#!/bin/sh
${huskyAppendIdentifier} start!
# v${version} ${platform}
hookName=\`basename "$0"\`
appendScriptPath="${script}.js"
selfPath="$(cd -P -- "$(dirname -- "$0")" && pwd -P)/$(basename -- "$0")"
${
  platform !== 'win32'
    ? `
if ! command -v node >/dev/null 2>&1; then
  echo "Can't find node in PATH, trying to find a node binary on your system"
fi
`
    : ''
}
if [ -f $appendScriptPath ]; then
  ${node} $appendScriptPath $selfPath $* || exit $?
else
  echo "Can't find user's $hookName hook"
fi
${huskyAppendIdentifier} end!
`

export function remove(data: string): string {
  const regx = new RegExp(
    `${huskyAppendIdentifier} start![^]+?${huskyAppendIdentifier} end!`,
    'g'
  )
  return data.replace(regx, '')
}

/**
 * @param rootDir - e.g. /home/typicode/project/
 * @param huskyDir - e.g. /home/typicode/project/node_modules/husky/
 * @param requireRunNodePath - path to run-node resolved by require e.g. /home/typicode/project/node_modules/.bin/run-node
 * @param platform - platform husky installer is running on (used to produce win32 specific script)
 */
export default function(
  rootDir: string,
  huskyDir: string,
  requireRunNodePath: string = require.resolve('.bin/run-node'),
  // Additional param used for testing only
  platform: string = os.platform()
) {
  const runNodePath = slash(path.relative(rootDir, requireRunNodePath))

  // On Windows do not rely on run-node
  const node = platform === 'win32' ? 'node' : runNodePath

  const { version } = JSON.parse(
    fs.readFileSync(path.join(__dirname, '../../package.json'), 'utf-8')
  )

  const script = slash(
    path.join(path.relative(rootDir, huskyDir), 'lib/runner/append-bin')
  )

  return render({ node, platform, script, version })
}
