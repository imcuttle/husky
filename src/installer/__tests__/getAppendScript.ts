import * as del from 'del'
import * as execa from 'execa'
import * as fs from 'fs'
import * as mkdirp from 'mkdirp'
import * as path from 'path'
import * as tempy from 'tempy'

import getAppendScript, {
  huskyAppendIdentifier,
  remove
} from '../getAppendScript'
import { getUserStagedFilename } from '../index'

const rootDir = '/home/typicode/project'
const huskyDir = '/home/typicode/project/node_modules/husky'

const curHuskyDir = path.join(__dirname, '../../..')

// On OS X/Linux runNodePath gets resolved to the following value
// In order to make the test deterministic on AppVeyor, the value is hardcoded
const runNodePath = '/home/typicode/project/node_modules/run-node/run-node'

let tempDir

function mkdir(dir: string) {
  mkdirp.sync(path.join(tempDir, dir))
}

function getFilename(name: string) {
  return path.join(tempDir, name)
}

function execSync(
  filename: string,
  argsOrOptions: string[] | object,
  options?: object
) {
  if (Array.isArray(argsOrOptions)) {
    return execa.sync(path.join(tempDir, filename), argsOrOptions, options)
  }
  return execa.sync(path.join(tempDir, filename), argsOrOptions)
}

function writeFile(filename: string, data: string) {
  fs.writeFileSync(path.join(tempDir, filename), data)
}

function writeExecFile(filename: string, data: string) {
  writeFile(filename, data)
  fs.chmodSync(path.join(tempDir, filename), parseInt('0755', 8))
}

function readFile(filename: string) {
  return fs.readFileSync(path.join(tempDir, filename), 'utf-8')
}

function exists(filename: string) {
  return fs.existsSync(path.join(tempDir, filename))
}

function readdir(filename: string) {
  return fs.readdirSync(path.join(tempDir, filename))
}

describe('getAppendScript', () => {
  beforeEach(() => {
    delete process.env.INIT_CWD
    tempDir = tempy.directory()
  })
  afterEach(() => del(tempDir, { force: true }))

  it('should match snapshot (OS X/Linux)', () => {
    const script = getAppendScript(rootDir, huskyDir, runNodePath, 'darwin')
    expect(script).toMatchSnapshot()
    expect(script).toMatch('run-node')
  })

  it('should match snapshot (Windows)', () => {
    const script = getAppendScript(rootDir, huskyDir, runNodePath, 'win32')

    expect(script).toMatchSnapshot()
    expect(script).not.toMatch('run-node')
  })

  it('should remove append wrapped', () => {
    expect(
      remove(
        [
          `${huskyAppendIdentifier} start!`,
          'asdasdsadasds',
          `${huskyAppendIdentifier} end!`
        ].join('\n')
      )
    ).toBe('')

    expect(
      remove(
        [
          ` ${huskyAppendIdentifier} start!`,
          'asdasdsadasds',
          `${huskyAppendIdentifier} end!?`
        ].join('\n')
      )
    ).toBe(' ?')

    expect(
      remove(
        [
          `${huskyAppendIdentifier} start!`,
          'asdasdsadasds',
          `${huskyAppendIdentifier} end!`,
          `${huskyAppendIdentifier} end!`
        ].join('\n')
      )
    ).toBe(`\n${huskyAppendIdentifier} end!`)

    expect(
      remove(
        [
          `${huskyAppendIdentifier} start!`,
          `${huskyAppendIdentifier} end!`,
          `${huskyAppendIdentifier} start!`,
          `${huskyAppendIdentifier} end!`
        ].join('\n')
      )
    ).toBe(`\n`)
  })

  it("should remove append wrapper when user's hook is not found", () => {
    const script = getAppendScript(tempDir, curHuskyDir)
    writeExecFile('script.sh', script)
    expect(readFile('script.sh')).toContain(huskyAppendIdentifier)
    console.log("readFile('script.sh')", readFile('script.sh'))
    execSync('script.sh', ['777', 'sss'], { stdio: 'inherit' })
    console.log("readFile('script.sh')", readFile('script.sh'))
    // Verify removing
    expect(readFile('script.sh')).not.toContain(huskyAppendIdentifier)
  })

  it('should run well', () => {
    const script = getAppendScript(tempDir, curHuskyDir)
    writeExecFile('script.sh', script)
    writeExecFile(
      getUserStagedFilename('script.sh'),
      `echo $* > ${getFilename('echoed')}`
    )
    execSync('script.sh', ['777', 'sss'], { stdio: 'inherit' })

    expect(readFile('script.sh')).toContain(huskyAppendIdentifier)
    console.log("readFile('echoed')", readFile('echoed'))
    expect(readFile('echoed')).toEqual('777 sss\n')
  })
})
