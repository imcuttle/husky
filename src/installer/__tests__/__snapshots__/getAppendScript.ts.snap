// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getAppendScript should match snapshot (OS X/Linux) 1`] = `
"#!/bin/sh
# husky-append start!
# v1.0.0-rc.13 darwin
hookName=\`basename \\"$0\\"\`
appendScriptPath=\\"node_modules/husky/lib/runner/append-bin.js\\"
selfPath=\\"$(cd -P -- \\"$(dirname -- \\"$0\\")\\" && pwd -P)/$(basename -- \\"$0\\")\\"

if ! command -v node >/dev/null 2>&1; then
  echo \\"Can't find node in PATH, trying to find a node binary on your system\\"
fi

if [ -f $appendScriptPath ]; then
  node_modules/run-node/run-node $appendScriptPath $selfPath $* || exit $?
else
  echo \\"Can't find user's $hookName hook\\"
fi
# husky-append end!
"
`;

exports[`getAppendScript should match snapshot (Windows) 1`] = `
"#!/bin/sh
# husky-append start!
# v1.0.0-rc.13 win32
hookName=\`basename \\"$0\\"\`
appendScriptPath=\\"node_modules/husky/lib/runner/append-bin.js\\"
selfPath=\\"$(cd -P -- \\"$(dirname -- \\"$0\\")\\" && pwd -P)/$(basename -- \\"$0\\")\\"

if [ -f $appendScriptPath ]; then
  node $appendScriptPath $selfPath $* || exit $?
else
  echo \\"Can't find user's $hookName hook\\"
fi
# husky-append end!
"
`;
